flocker/node/agents/cinder.py

```diff
@@ -579,29 +579,42 @@ def get_device_path(self, blockdevice_id):
         # http://libvirt.org/formatdomain.html#elementsDisks (target section)
         # However, the correct device is named as a udev symlink which includes
         # the first 20 characters of the blockedevice_id.
-        device_path = FilePath(
-            "/dev/disk/by-id/virtio-{}".format(blockdevice_id[:20]))
-        if not device_path.exists():
-            # If the device path does not exist, either virtio driver is
-            # not being used (e.g. Rackspace), or the user has modified
-            # their udev rules.  The following code relies on Cinder
-            # returning the correct device path, which appears to work
-            # for Rackspace and will work with virtio if no disks have
-            # been attached outside Cinder.
-            try:
-                cinder_volume = self.cinder_volume_manager.get(blockdevice_id)
-            except CinderNotFound:
-                raise UnknownVolume(blockdevice_id)
+        expected_path = FilePath(
+            "/dev/disk/by-id/virtio-{}".format(volume.id[:20])
+        )
+        if expected_path.exists():
+            return expected_path
+        else:
+            raise UnattachedVolume(volume.id)
+
+    def _get_device_path_other(self, volume):
+        # If the device path does not exist, either virtio driver is
+        # not being used (e.g. Rackspace), or the user has modified
+        # their udev rules.  The following code relies on Cinder
+        # returning the correct device path, which appears to work
+        # for Rackspace and will work with virtio if no disks have
+        # been attached outside Cinder.
+
+        # As far as we know you can not have more than one attachment,
+        # but, perhaps we're wrong and there should be a test for the
+        # multiple attachment case.  FLOC-1854.
+        try:
+            [attachment] = volume.attachments
+        except ValueError:
+            raise UnattachedVolume(volume.id)
 
-            # As far as we know you can not have more than one attachment,
-            # but, perhaps we're wrong and there should be a test for the
-            # multiple attachment case.  FLOC-1854.
-            try:
-                [attachment] = cinder_volume.attachments
-            except ValueError:
-                raise UnattachedVolume(blockdevice_id)
+        return FilePath(attachment['device'])
 
-            device_path = FilePath(attachment['device'])
+    def get_device_path(self, blockdevice_id):
+        try:
+            cinder_volume = self.cinder_volume_manager.get(blockdevice_id)
+        except CinderNotFound:
+            raise UnknownVolume(blockdevice_id)
+
+        if FilePath('/sys/bus/virtio/drivers/virtio_blk').isdir():
```

Comment:
:question: wrap this in a simple `is_virtio` function to clarify the decision being made.  The docstring for this simple function should explain how the existence of the this directory shows which driver is in operation.